base/shell/explorer/notifyiconscust.cpp

```diff
@@ -20,31 +20,193 @@
 
 #include "precomp.h"
 
+#define BITMAP_WIDTH 16
+#define BITMAP_HEIGHT 16
+
+#define IDC_NOTIFY_OPTIONS 5000
+
+typedef struct
+{
+    DWORD hwnd;
+    DWORD uId;
+    DWORD uCallbackMsg;
+    DWORD unkow[2];
+    HICON hIcon;
+    WCHAR text[64];
+}TRAYDATA;
+
+// Returns TRUE if successful, and FALSE otherwise. 
+BOOL AddListViewColumn(HWND hWndListView, LPWSTR label, INT iCol, int w)
+{
+    LVCOLUMNW lvc;
+
+    lvc.mask = LVCF_FMT | LVCF_WIDTH | LVCF_TEXT | LVCF_SUBITEM;
+    lvc.iSubItem = iCol;
+    lvc.pszText = label;
+    lvc.cchTextMax = lstrlenW(label);
+    lvc.cx = w;
+    lvc.fmt = LVCFMT_CENTER;
+
+    return (ListView_InsertColumn(hWndListView, iCol, &lvc) != -1);
+}
+
+// Returns TRUE if successful, and FALSE otherwise. 
+BOOL InsertItem(HWND hListView, TBBUTTON *tb, TRAYDATA *data, int id)
+{
+    LVITEMW iItem;
+
+    iItem.mask = LVIF_TEXT | LVIF_IMAGE;
+    iItem.pszText = data->text;
+    iItem.cchTextMax = lstrlenW(data->text);
+    iItem.iSubItem = 0;
+    iItem.iItem = id;
+    iItem.iImage = id;
+
+    return (ListView_InsertItem(hListView, &iItem) != -1);
+}
+
+// Return Toolbar Window hander on successful, Null otherwise
+HWND FindTrayToolbarWindowHandle()
+{
+    HWND hwnd;
+    if ((hwnd = FindWindowW(L"Shell_TrayWnd", NULL)))
+        if ((hwnd = FindWindowExW(hwnd, NULL, L"TrayNotifyWnd", NULL)))
+            if ((hwnd = FindWindowExW(hwnd, NULL, L"SysPager", NULL)))
+                if ((hwnd = FindWindowExW(hwnd, NULL, L"ToolbarWindow32", NULL)))
+                    return hwnd;
+
+    return NULL;
+}
+
+BOOL ExtractData(HWND hListView, HIMAGELIST hImageList)
+{
+    const int BUFFER_SIZE = 0x1000;
+    VOID* pLocalBuffer = (VOID*)new CHAR[BUFFER_SIZE];
+
+    HWND hToolbar = FindTrayToolbarWindowHandle();
+    int count = SendMessageW(hToolbar, TB_BUTTONCOUNT, 0, 0);
+
+    DWORD processId = 0;
+    GetWindowThreadProcessId(hToolbar, &processId);
+    HANDLE hProcess = OpenProcess(PROCESS_ALL_ACCESS, FALSE, processId);
+    if (!hProcess)
+        return FALSE;
+
+    VOID *pRemoteBuffer = VirtualAllocEx(hProcess, NULL, BUFFER_SIZE, MEM_COMMIT, PAGE_READWRITE);
+    if (!pRemoteBuffer)
+        return FALSE;
+
+    ZeroMemory(pLocalBuffer, BUFFER_SIZE);
+
+    hImageList = ImageList_Create(BITMAP_WIDTH, BITMAP_HEIGHT, FALSE, count, 0);
```

Comment:
what if hImageList is NULL ?

Reply:
Just don't show the icons, everything else works normal.